File: src/Windows/Data/UsageSnapshot.cs
namespace ClaudeStatusBar.Data;

/// <summary>
/// A single parsed get_usage response. Utilization values are the raw 0-100 percent
/// the protocol reports (49, 69.0, ...) -- NOT 0-1 fractions; callers building a
/// GaugeModel divide by 100. WeeklyUtilization/WeeklyResetsAt are null when the
/// "all models" weekly node could not be located (protocol drift) -- the ring then
/// simply renders empty rather than the app crashing or restart-looping.
///
/// SubscriptionType is the plan (max, team, ...) this response's account is on
/// (docs/multi-account.md), carried through purely for account labelling
/// (Model/AccountLabel.cs) -- it is deliberately NOT part of QuotaModel's ingest
/// math or QuotaView, so the model contract stays unchanged. Defaulted so every
/// existing positional `new UsageSnapshot(...)` call site keeps compiling.
///
/// SessionIsActive/WeeklyIsActive: whether the server says this window is
/// currently open (docs/forecast-and-states.md, "A closed window is an answer,
/// not an absence"). A 5 h session that has simply expired -- or a weekly window
/// just past its reset with nothing consumed -- comes back as {utilization: 0,
/// resets_at: null, is_active: false}: a real, meaningful state, not a broken
/// response. Default true so every existing positional call site (which never
/// mentions this field, and always supplies a resets_at when it means "open")
/// keeps compiling and keeps its old meaning.
/// </summary>
public sealed record UsageSnapshot(
    double SessionUtilization,
    string? SessionResetsAt,
    double? WeeklyUtilization,
    string? WeeklyResetsAt,
    DateTimeOffset ReceivedAt,
    string? SubscriptionType = null,
    bool SessionIsActive = true,
    bool WeeklyIsActive = true);

File: src/Mac/Sources/ClaudeQuotaCore/UsageSnapshot.swift
import Foundation

/// One `get_usage` reply, reduced to what the model needs. Mirrors
/// `src/Windows/Data/UsageSnapshot.cs`.
///
/// `sessionResetsAt` and `weeklyResetsAt` are the **raw strings** from the wire, not parsed
/// dates, because the exact byte sequence is the snapshot fingerprint the model dedups on
/// (docs/forecast-and-states.md, "Ingest" rule 2). Parsing and re-formatting them would destroy
/// the microsecond digits that carry the signal.
public struct UsageSnapshot: Equatable, Sendable {
    public let sessionUtilization: Double?
    public let sessionResetsAt: String?
    public let weeklyUtilization: Double?
    public let weeklyResetsAt: String?
    public let subscriptionType: String?
    /// Whether the server says this window is currently open.
    ///
    /// A 5 h session that has simply expired comes back as
    /// `{"utilization": 0, "resets_at": null, "is_active": false}`. That is a real, meaningful
    /// state — "nothing running, nothing consumed" — **not** a broken response. An earlier cut of
    /// the parser required `resets_at` and therefore rejected the whole reading, which put the
    /// menu-bar icon into "cannot read the quota" for five hours on a perfectly healthy account.
    public let sessionIsActive: Bool
    public let weeklyIsActive: Bool
    public let observedAt: Date
    /// Which reading strategy produced this snapshot -- recorded so a regression that silently
    /// falls back to the fragile path is visible rather than invisible. See `UsageParser`.
    public let source: Source

    public enum Source: String, Sendable {
        case limitsArray = "limits[]"
        case exactKeys = "exact-keys"
        case substringSearch = "substring-search"
    }

    public init(
        sessionUtilization: Double?,
        sessionResetsAt: String?,
        weeklyUtilization: Double?,
        weeklyResetsAt: String?,
        subscriptionType: String?,
        sessionIsActive: Bool = true,
        weeklyIsActive: Bool = true,
        observedAt: Date,
        source: Source
    ) {
        self.sessionUtilization = sessionUtilization
        self.sessionResetsAt = sessionResetsAt
        self.weeklyUtilization = weeklyUtilization
        self.weeklyResetsAt = weeklyResetsAt
        self.subscriptionType = subscriptionType
        self.sessionIsActive = sessionIsActive
        self.weeklyIsActive = weeklyIsActive
        self.observedAt = observedAt
        self.source = source
    }

    /// docs/forecast-and-states.md, round-1 decision 9: utilization must be finite and in
    /// [0, 100] at the boundary, or the window is treated as missing rather than as 0 %.
    public static func isValidPercentage(_ value: Double?) -> Bool {
        guard let value else { return false }
        return value.isFinite && value >= 0 && value <= 100
    }

    /// A usable percentage. A *reset* is not required: an inactive window legitimately has none.
    public var hasValidSession: Bool { Self.isValidPercentage(sessionUtilization) }
    public var hasValidWeekly: Bool { Self.isValidPercentage(weeklyUtilization) }

    /// What the forecast model needs: a percentage AND a deadline to measure against. Without a
    /// deadline the window can only be reported, never forecast — which is what
    /// docs/forecast-and-states.md's refusal rules already say.
    public var sessionIsForecastable: Bool { hasValidSession && sessionResetsAt != nil }
    public var weeklyIsForecastable: Bool { hasValidWeekly && weeklyResetsAt != nil }
}
